src/daft-functions/src/binary/codecs.rs

```diff
@@ -0,0 +1,142 @@
+use std::borrow::Cow;
+
+use common_error::{DaftError, DaftResult};
+use serde::{Deserialize, Serialize};
+
+/// Supported codecs for the decode and encode functions.
+#[derive(Debug, Clone, Copy, Serialize, Deserialize, PartialEq, Eq, Hash)]
+pub enum Codec {
+    Gzip,
+    Zlib,
+    Deflate,
+}
+
+/// Function type for encoding a string to bytes
+pub(crate) type Encoder = fn(input: &str) -> DaftResult<Cow<'_, [u8]>>;
+
+/// Function type for decoding bytes to a string
+pub(crate) type Decoder = fn(input: &[u8]) -> DaftResult<Cow<'_, str>>;
+
+/// Each codec should have an encode/decode pair.
+impl Codec {
+    pub(crate) fn encoder(&self) -> Encoder {
+        match self {
+            Self::Gzip => gzip_encoder,
+            Self::Zlib => zlib_encoder,
+            Self::Deflate => deflate_encoder,
+        }
+    }
+
+    pub(crate) fn decoder(&self) -> Decoder {
+        match self {
+            Self::Deflate => deflate_decoder,
+            Self::Gzip => gzip_decoder,
+            Self::Zlib => zlib_decoder,
+        }
+    }
+}
+
+impl TryFrom<&str> for Codec {
+    type Error = DaftError;
+
+    fn try_from(s: &str) -> Result<Self, Self::Error> {
+        match s.to_lowercase().as_str() {
+            "deflate" => Ok(Self::Deflate),
+            "gzip" => Ok(Self::Gzip),
+            "zlib" => Ok(Self::Zlib),
+            _ => Err(DaftError::not_implemented(format!(
+                "unsupported codec: {}",
+                s
+            ))),
+        }
+    }
+}
+
+//
+// ENCODERS
+//
+
+#[inline]
+fn deflate_encoder(input: &str) -> DaftResult<Cow<'_, [u8]>> {
+    use std::io::Write;
+
+    use flate2::{write::DeflateEncoder, Compression};
+    let mut encoder = DeflateEncoder::new(Vec::new(), Compression::default());
+    encoder.write_all(input.as_bytes())?;
+    Ok(encoder.finish()?.into())
+}
+
+#[inline]
+fn gzip_encoder(input: &str) -> DaftResult<Cow<'_, [u8]>> {
+    use std::io::Write;
+
+    use flate2::{write::GzEncoder, Compression};
+    let mut encoder = GzEncoder::new(Vec::new(), Compression::default());
+    encoder.write_all(input.as_bytes())?;
+    Ok(encoder.finish()?.into())
+}
+
+#[inline]
+fn zlib_encoder(input: &str) -> DaftResult<Cow<'_, [u8]>> {
+    use std::io::Write;
+
+    use flate2::{write::ZlibEncoder, Compression};
+    let mut encoder = ZlibEncoder::new(Vec::new(), Compression::default());
+    encoder.write_all(input.as_bytes())?;
+    Ok(encoder.finish()?.into())
+}
+
+//
+// DECODERS
+//
+
+#[inline]
+fn deflate_decoder(input: &[u8]) -> DaftResult<Cow<'_, str>> {
```

Comment:
same here. are we ever producing a `Cow<'str>` or is it always an owned `String`? It looks like it's always the latter.